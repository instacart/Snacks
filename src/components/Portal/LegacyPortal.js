import React, { Component }   from 'react'
import ReactDOM               from 'react-dom'
import PropTypes              from 'prop-types'

const DEFAULT_CONTAINER = document.body

class Portal extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    container: PropTypes.element
  }

  constructor () {
    super()
    this.setContainer(this.props.container)
  }

  componentDidMount() {
    this.renderPortal()
  }

  componentDidUpdate() {
    this.renderPortal()
  }

  componentWillUnmount() {
    ReactDOM.unmountComponentAtNode(this.containerEl)
    this.containerEl.parent.removeChild(this.containerEl)
    this.containerEl = null
  }

  setContainer (nextContainer) {
    if (!nextContainer) {
      this.containerEl = document.createElement('div')
      DEFAULT_CONTAINER.appendChild(this.containerEl)
    } else {
      this.containerEl = nextContainer
    }
  }

  renderPortal() {
    let children = this.props.children
    if (typeof this.props.children.type === 'function') {
      children = React.cloneElement(this.props.children)
    }

    ReactDOM.unstable_renderSubtreeIntoContainer(
      this,
      children,
      this.containerEl
    )
  }

  render() {
    return null
  }
}

export default Portal
