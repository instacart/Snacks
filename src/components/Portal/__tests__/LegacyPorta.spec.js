import React from 'react'
import renderer from 'react-test-renderer'
import { StyleRoot } from 'radium'
import { mount } from 'enzyme'
import LegacyPortal from '../LegacyPortal'

describe('Legacy Portal', () => {
  it('renders properly', () => {
    const tree = renderer
      .create(
        <StyleRoot>
          <LegacyPortal>
            <p>Test content</p>
          </LegacyPortal>
        </StyleRoot>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders with a passed in conatiner', () => {
    const container = mount(<div />)
    const Portal = mount(
      <StyleRoot>
        <LegacyPortal container={container.instance()}>{<p>Test content</p>}</LegacyPortal>
      </StyleRoot>
    )

    expect(Portal).toBeTruthy()
    expect(Portal.find(LegacyPortal).props().container).toEqual(container.instance())
  })
})
