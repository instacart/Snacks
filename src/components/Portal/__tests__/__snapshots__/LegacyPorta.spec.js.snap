// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legacy Portal renders properly 1`] = `null`;
